bin/pkg/lock.ml

```diff
@@ -119,19 +122,29 @@ let solve_lock_dir
     ~constraints:(constraints_of_workspace workspace ~lock_dir_path)
   >>= function
   | Error (`Diagnostic_message message) -> Fiber.return (Error (lock_dir_path, message))
-  | Ok { lock_dir; files; pinned_packages } ->
+  | Ok { lock_dir; files; pinned_packages; num_expanded_packages } ->
+    let time_end = Unix.gettimeofday () in
     let summary_message =
       User_message.make
-        [ Pp.tag
-            User_message.Style.Success
-            (Pp.textf
-               "Solution for %s:"
-               (Path.Source.to_string_maybe_quoted lock_dir_path))
-        ; (match Package_name.Map.values lock_dir.packages with
-           | [] ->
-             Pp.tag User_message.Style.Warning @@ Pp.text "(no dependencies to lock)"
-           | packages -> pp_packages packages)
-        ]
+        ([ Pp.tag
+             User_message.Style.Success
+             (Pp.textf
+                "Solution for %s:"
+                (Path.Source.to_string_maybe_quoted lock_dir_path))
+         ; (match Package_name.Map.values lock_dir.packages with
+            | [] ->
+              Pp.tag User_message.Style.Warning @@ Pp.text "(no dependencies to lock)"
+            | packages -> pp_packages packages)
+         ]
+         @
+         if print_perf_stats
+         then
+           [ Pp.nop
+           ; Pp.textf "Expanded packages: %d" num_expanded_packages
+           ; Pp.textf "Updated repos in: %.2fs" (time_solve_start -. time_start)
+           ; Pp.textf "Solved dependencies in: %.2fs" (time_end -. time_solve_start)
+           ]
+         else [])
```

Comment:
Could you unwrap this a little, e.g. by defining the perf stats as a separate binding (containint the stats or `[]`), then you can always cons the actual message easily and don't need quadratic `@`.

This way the code does a lot of stuff, wrapped deeply and I find it a bit hard to read, as it requires more and more context the more layers are added.

Reply:
Good idea. Done.